### DISCUSSION TITLE
Jitting models

### DISCUSSION THREAD

**Discussion by jithendaraa** [google/flax#1728] [Q&A]
Hi. What's the right flax linen way to use JAX's JIT while playing with neural networks? It's a bit unclear to me when one would use `jax.jit` decorated methods inside the network Class vs. `flax.linen.jit`.

**Comment by jheek** [comment]
Inside a linen Module you should always use the wrappers provided by flax.
For jit in particular though I would really try to push it outside of flax. It's much easier to get good performance when you can jit the entire model call (with jax.jit). So typically you just want to use at least `jax.jit(module.apply)(...)`  or wrap a function in jit that includes the module.apply

**Reply by jithendaraa** [reply]
Hmmm, I tried doing that but i get `TypeError: Failed to hash Flax Module.  The module probably contains unhashable attributes`. Why are these unhashable created and how do I sidestep the issue? My Flax Model consists of many functions, not just `__call__`

**Reply by jheek** [reply]
You could manually add a `__hash__(self): return id(self)`. Ideally the default hash just works of course. Perhaps you added numpy arrays as attribute values or something else that is not hashable?
